cranelift/peepmatic/crates/souper/src/lib.rs

```diff
@@ -0,0 +1,711 @@
+/// Maximum recursion depth, to avoid blowing the stack.
+const MAX_DEPTH: u8 = 50;
+
+/// Convert a file containing Souper optimizations into Peepmatic DSL.
+pub fn convert_file(path: &Path) -> Result<String> {
+    let souper = std::fs::read_to_string(path)
+        .with_context(|| format!("failed to read: {}", path.display()))?;
+    convert_str(&souper, Some(path))
+}
+
+/// Convert a string of Souper optimizations into Peepmatic DSL.
+///
+/// The optional `filename` parameter is used for better error messages.
+pub fn convert_str(souper: &str, filename: Option<&Path>) -> Result<String> {
+    let mut peepmatic = String::new();
+
+    let replacements = souper_ir::parse::parse_replacements_str(souper, filename)?;
+    for replacement in replacements {
+        let (statements, lhs, rhs) = match replacement {
+            ast::Replacement::LhsRhs {
+                statements,
+                lhs,
+                rhs,
+            } => {
+                if !lhs.attributes.is_empty() {
+                    log::warn!("cannot translate Souper attributes to Peepmatic DSL");
+                    continue;
+                }
+                (statements, lhs.value, rhs)
+            }
+            ast::Replacement::Cand { statements, cand } => {
+                if !cand.attributes.is_empty() {
+                    log::warn!("cannot translate Souper attributes to Peepmatic DSL");
+                    continue;
+                }
+                let lhs = match cand.lhs {
+                    ast::Operand::Value(v) => v,
+                    ast::Operand::Constant(_) => {
+                        log::warn!("optimization's LHS must not be a constant");
+                        continue;
+                    }
+                };
+                (statements, lhs, cand.rhs)
+            }
+        };
+
+        if let Some(s) = convert_replacement(&statements, lhs, rhs) {
+            peepmatic.push_str(&s);
+            peepmatic.push('\n');
+        }
+    }
+
+    Ok(peepmatic)
+}
+
+fn convert_replacement(
+    statements: &ast::Arena<ast::Statement>,
+    lhs: ast::ValueId,
+    rhs: ast::Operand,
+) -> Option<String> {
+    let lhs = convert_lhs(statements, lhs)?;
+    let rhs = convert_rhs(statements, rhs)?;
+    Some(format!("(=> {}\n    {})\n", lhs, rhs))
+}
+
+fn convert_lhs(statements: &ast::Arena<ast::Statement>, lhs: ast::ValueId) -> Option<String> {
+    let mut tys = vec![];
+    let pattern = convert_operand(statements, lhs.into(), &mut tys, 0)?;
+
+    Some(if tys.is_empty() {
+        pattern
+    } else {
+        let mut lhs = format!("(when {}", pattern);
+        for (name, width) in tys {
+            lhs.push_str("\n        ");
+            lhs.push_str(&format!("(bit-width ${} {})", name, width));
+        }
+        lhs.push(')');
+        lhs
+    })
+}
+
+fn convert_name(name: &str) -> String {
+    debug_assert!(name.starts_with('%'));
+    debug_assert!(name.len() >= 2);
+    let c = name.chars().nth(1).unwrap();
```

Comment:
You could use `as_bytes()[1]` as `%` is in the ascii range and you are checking if it is an ascii lowercase alphabetic character.